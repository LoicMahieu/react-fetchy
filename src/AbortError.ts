
export class AbortError extends Error {
  public message = "Request has been aborted";
}
