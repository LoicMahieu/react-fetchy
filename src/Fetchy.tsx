import { isEqual, pick } from "lodash";
import * as PropTypes from "prop-types";
import * as React from "react";
import * as request from "superagent";

export interface IState {
  fulfilled: boolean;
  pending: boolean;
  rejected: boolean;

  error: any;
  progress: any;
  result: any;
  value: any;
}

type methodType = "get" | "post" | "put" | "delete";

export interface IOptions {
  body?: any;
  method?: methodType;
  url?: string;
  query?: object;
  headers?: object;
  then?(state: IState): Promise<any>;
}

export interface IBag {
  state: IState;
  fetch(options: IOptions): Promise<IState>;
  reset(): void;
}

interface IProps extends IOptions {
  children?(bag: IBag): React.ReactNode;
  render?(bag: IBag): React.ReactNode;
}

const initialState: IState = {
  fulfilled: false,
  pending: false,
  rejected: false,

  error: null,
  progress: null,
  result: null,
  value: null,
};

const compareProps = [
  "url",
  "query",
  "headers",
  "method",
];

export default class Fetchy extends React.Component<IProps, IState> {
  public static defaultProps = {};

  public static propType = {
    title: PropTypes.string,
  };

  public readonly state: IState = initialState;

  private req: any;

  public async componentDidMount() {
    if (this.props.url) {
      this.fetch();
    }
  }

  public componentDidUpdate(prevProps: IProps) {
    if (!isEqual(pick(this.props, compareProps), pick(prevProps, compareProps))) {
      this.fetch();
    }
  }

  public componentWillUnmount() {
    if (this.req) {
      this.req.abort();
    }
  }

  public render() {
    const render = typeof this.props.children === "function"
      ? this.props.children
      : this.props.render;

    if (!render) {
      throw new Error("Missing children or render props");
    }

    const bag: IBag = {
      fetch: this.fetch,
      reset: this.reset,
      state: Object.freeze({ ...this.state }),
    };

    return render(bag);
  }

  private fetch = async (options: IOptions = {}) => {
    if (this.req) {
      this.req.abort();
    }

    const mergedOptions = {
      ...this.props,
      ...options,
    };

    const {
      url,
      method = "get",
      query,
      body,
      headers,
      then,
    } = mergedOptions;

    if (!url) {
      throw new Error("Parameter `url` is required.");
    }

    this.setState({ pending: true, fulfilled: false });

    let response: any;

    // Create request
    const m = method.toLowerCase() as methodType;
    const createReq = request[m];
    if (!createReq)Â {
      throw new Error(`Invalid method ${method}.`);
    }
    const req = createReq(url);
    this.req = req;

    if (headers) {
      req.set(headers);
    }
    if (query) {
      req.query(query);
    }
    if (body) {
      req.send(body);
    }
    req.on("progress", (progress) => {
      this.setState((oldState) => ({ ...oldState, progress }));
    });
    req.on("response", (incomingMessage) => {
      response = incomingMessage;
    });

    try {
      // Wait response
      await req;

      /* istanbul ignore if : should never happend */
      if (!response) {
        throw new Error("No response");
      }

      // Handle Result
      const value = response.body;
      const state: IState = {
        ...initialState,
        fulfilled: true,
        result: response,
        value,
      };
      if (then) {
        Object.assign(state, then(state));
      }
      this.setState(state);
    } catch (error) {
      error.data = error.data || {};
      error.data.result = {
        response,
      };

      this.setState({
        error,
        fulfilled: false,
        pending: false,
        rejected: true,
        result: response,
        value: null,
      });

      throw error;
    }

    return Object.freeze({ ...this.state });
  }

  private reset = () => {
    this.setState({ ...initialState });
  }
}
