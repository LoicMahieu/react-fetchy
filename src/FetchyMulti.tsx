// throw error if no options.id
// check error on timeout

// TODOS:
// * Pause/resume

import * as React from "react";
import * as request from "superagent";
import {
  IFetchyRequestOptions,
  IFetchyState,
  IMethodType,
  initialState as singleInitialState,
} from "./Fetchy";

export interface IFetchyMultiState {
  [id: string]: IFetchyState | undefined;
}

export interface IFetchyMultiRequestOptions extends IFetchyRequestOptions {
  id: string;
}

export interface IFetchyMultiOptions {
  concurrency: number;
  requests: IFetchyMultiRequestOptions[];
}

export interface IFetchyMultiRenderArgs {
  states: {
    [id: string]: IFetchyState;
  };
  abort: (id?: string) => void;
}

interface IFetchyMultiProps extends IFetchyMultiOptions {
  children?(bag: IFetchyMultiRenderArgs): React.ReactNode;
  render?(bag: IFetchyMultiRenderArgs): React.ReactNode;
}

const initialState: IFetchyMultiState = {};

class AbortError extends Error {
  public message = "Request has been aborted";
}

export class FetchyMulti extends React.Component<IFetchyMultiProps, IFetchyMultiState> {
  public static defaultProps: IFetchyMultiProps = {
    concurrency: 2,
    requests: [],
  };

  public readonly state: IFetchyMultiState = initialState;

  private superAgentRequests: {
    [id: string]: request.SuperAgentRequest;
  } = {};

  public async componentDidMount() {
    this.checkQueue();
  }

  public componentDidUpdate(previousProps: IFetchyMultiProps) {
    if (previousProps === this.props) {
      return;
    }

    // Clean state
    this.setState(
      state =>
        Object.keys(state).reduce((newState, id) => {
          const request = this.props.requests.find(
            request => request.id === id,
          );
          return {
            ...newState,
            [id]: request ? state[id] : undefined,
          };
        }, {}),
      () => {
        // Abort running
        this.superAgentRequests = Object.keys(this.superAgentRequests).reduce((newRequests, id) => {
          const request = this.props.requests.find(
            request => request.id === id,
          );

          if (request) {
            return {
              ...newRequests,
              [id]: this.superAgentRequests[id],
            };
          } else {
            this.abort(id);
            return newRequests;
          }
        }, {});

        this.checkQueue();
      },
    );
  }

  public componentWillUnmount() {
    this.abort();
  }

  public render() {
    const render =
      typeof this.props.children === "function"
        ? this.props.children
        : this.props.render;

    if (!render) {
      throw new Error("Missing children or render props");
    }

    const bag: IFetchyMultiRenderArgs = {
      abort: this.abort,
      states: this.props.requests.reduce(
        (bagStates, request) => ({
          ...bagStates,
          [request.id]: this.state[request.id] || singleInitialState,
        }),
        {},
      ),
    };

    return render(bag);
  }

  private fetch = async (options: IFetchyMultiRequestOptions) => {
    if (this.superAgentRequests[options.id]) {
      this.superAgentRequests[options.id].abort();
      delete this.superAgentRequests[options.id];
    }

    const {
      url,
      method = "get",
      query,
      body,
      headers,
      then,
      timeout,
      retry,
      withCredentials,
    } = options;

    if (!url) {
      throw new Error("Parameter `url` is required.");
    }

    this.setState({
      [options.id]: {
        ...singleInitialState,
        fulfilled: false,
        pending: true,
      },
    });

    let response: any;

    // Create request
    const m = method.toLowerCase() as IMethodType;
    const createReq = request[m];
    if (!createReq) {
      throw new Error(`Invalid method ${method}.`);
    }
    const req = createReq(url);
    this.superAgentRequests[options.id] = req;

    if (headers) {
      req.set(headers);
    }
    if (query) {
      req.query(query);
    }
    if (body) {
      req.send(body);
    }
    if (timeout) {
      req.timeout(timeout);
    }
    if (retry) {
      req.retry(retry);
    }
    if (withCredentials) {
      req.withCredentials();
    }
    req.on("progress", progress => {
      this.setState(state => ({
        [options.id]: {
          ...(state[options.id] || singleInitialState),
          progress,
        },
      }));
    });
    req.on("response", incomingMessage => {
      response = incomingMessage;
    });

    try {
      // Wait response
      await req;

      /* istanbul ignore if : should never happend */
      if (!response) {
        throw new Error("No response");
      }

      // Handle Result
      const value = response.body;
      const state: IFetchyState = {
        ...singleInitialState,
        fulfilled: true,
        result: response,
        value,
      };
      if (then) {
        Object.assign(state, then(state));
      }
      this.setState({
        [options.id]: state,
      });
    } catch (error) {
      console.log("errror ?");

      error.data = error.data || {};
      error.data.result = {
        response,
      };

      this.setState({
        [options.id]: {
          error,
          fulfilled: false,
          pending: false,
          rejected: true,
          result: response,
          value: null,
        },
      });

      throw error;
    }

    return Object.freeze({ ...this.state });
  };

  private async checkQueue() {
    const { concurrency } = this.props;
    const runningRequestCount = Object.keys(this.state).filter(id => {
      const state = this.state[id];
      return state ? state.pending : false;
    }).length;
    const availableCount = concurrency - runningRequestCount;
    const requestsTodo = this.props.requests.filter(request => {
      const runningRequest = this.superAgentRequests[request.id];
      if (runningRequest) {
        return false;
      }
      return true;
    });

    await Promise.all(
      requestsTodo.slice(0, availableCount).map(async request => {
        try {
          await this.fetch(request);
        } finally {
          this.checkQueue();
        }
      }),
    );
  }

  private abort = (id?: string) => {
    if (id) {
      if (this.superAgentRequests[id]) {
        this.superAgentRequests[id].abort();
      }
      this.setState({
        [id]: {
          error: new AbortError(),
          fulfilled: false,
          pending: false,
          rejected: true,
          result: undefined,
          value: null,
        },
      });
    } else {
      const ids = [
        ...Object.keys(this.superAgentRequests),
        ...this.props.requests.map(req => req.id),
      ];
      ids.map(id => {
        this.abort(id);
      });
    }
  };
}
