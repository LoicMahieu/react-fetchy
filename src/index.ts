export * from "./Fetchy";
export * from "./FetchyMulti";
