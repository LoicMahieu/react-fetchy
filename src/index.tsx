
import Fetchy from './Fetchy'
export default Fetchy
export {
  Fetchy
}
