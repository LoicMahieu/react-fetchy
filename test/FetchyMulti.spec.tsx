import waitUntil from "async-wait-until";
import { configure, shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import * as React from "react";

import { FetchyMulti, IFetchyMultiRenderArgs } from "../src";
import { beforeRunHttpServer, setup } from "./utils";

const base = "http://localhost:8081";

configure({ adapter: new Adapter() });

describe("FetchyMulti", () => {
  beforeRunHttpServer(8081);

  it("no render or children", () => {
    expect(() => shallow(<FetchyMulti />)).toThrow(
      "Missing children or render props",
    );
  });
  describe("basic", () => {
    it("render", () => {
      const render = jest.fn((bag: IFetchyMultiRenderArgs) => {
        expect(bag).toMatchSnapshot();
      });
      shallow(<FetchyMulti render={render} />);
      expect(render).toHaveBeenCalled();
    });
    it("children", () => {
      const children = jest.fn((bag: IFetchyMultiRenderArgs) => {
        expect(bag).toMatchSnapshot();
      });
      shallow(<FetchyMulti children={children} />);
      expect(children).toHaveBeenCalled();
    });
  });
  it("fetch on mount", async () => {
    const then = jest.fn();
    const res = setup(FetchyMulti, {
      requests: [
        {
          id: "1",

          headers: {
            "Some-Header": "1",
          },
          query: "foo",
          retry: 5,
          timeout: 20,
          url: `${base}/200`,
          withCredentials: true,

          then,
        },
        {
          id: "2",
          url: `${base}/200?delay=50`,
        },
      ],
    });
    expect(res.bag.states["1"].pending).toEqual(true);
    expect(res.bag.states["2"].pending).toEqual(true);

    await waitUntil(() => !res.bag.states["1"].pending);
    expect(res.bag.states["1"].fulfilled).toEqual(true);

    expect(then).toHaveBeenCalled();

    await waitUntil(() => !res.bag.states["2"].pending);
    expect(res.bag.states["2"].fulfilled).toEqual(true);

    expect(res.bag).toMatchSnapshot();
  });
  it("abort on unmount", async () => {
    const res = setup(FetchyMulti, {
      requests: [
        {
          id: "1",
          url: `${base}/200`,
        },
        {
          id: "2",
          url: `${base}/200?delay=100`,
        },
      ],
    });
    expect(res.bag.states["1"].pending).toEqual(true);
    expect(res.bag.states["2"].pending).toEqual(true);

    res.wrapper.unmount();

    expect(res.bag.states["1"].rejected).toEqual(true);
    expect(res.bag.states["2"].rejected).toEqual(true);
  });
  it("handle errors", async () => {
    const res = setup(FetchyMulti, {
      requests: [
        {
          id: "1",
          url: `${base}/500`,
        },
      ],
    });
    expect(res.bag.states["1"].pending).toEqual(true);
    await waitUntil(() => !res.bag.states["1"].pending);

    expect(res.bag.states["1"].fulfilled).toEqual(false);
    expect(res.bag.states["1"].rejected).toEqual(true);

    expect(res.bag).toMatchSnapshot();
  });
  it("handle errors and retry", async () => {
    const res = setup(FetchyMulti, {
      requests: [
        {
          id: "1",
          url: `${base}/500`,
        },
      ],
    });
    expect(res.bag.states["1"].pending).toEqual(true);
    await waitUntil(() => !res.bag.states["1"].pending);

    expect(res.bag.states["1"].fulfilled).toEqual(false);
    expect(res.bag.states["1"].rejected).toEqual(true);

    expect(res.bag).toMatchSnapshot();

    expect(() => res.bag.retry("123")).toThrowError(
      "Could not find request for id: 123",
    );

    res.bag.retry("1");
    expect(res.bag.states["1"].pending).toEqual(true);
  });
  it("abort on update", async () => {
    const res = setup(FetchyMulti, {
      requests: [
        {
          id: "1",
          url: `${base}/200`,
        },
        {
          id: "2",
          url: `${base}/200`,
        },
      ],
    });
    expect(res.bag.states["1"].pending).toEqual(true);
    expect(res.bag.states["2"].pending).toEqual(true);

    res.wrapper.setProps({
      requests: [
        {
          id: "2",
          url: `${base}/200`,
        },
        {
          id: "3",
          url: `${base}/200`,
        },
      ],
    });

    expect(res.bag.states["1"]).toBeUndefined();
    expect(res.bag.states["2"].pending).toEqual(true);
    expect(res.bag.states["3"].pending).toEqual(true);
  });
  it("abort not started requests", async () => {
    const res = setup(FetchyMulti, {
      requests: [
        {
          id: "1",
          url: `${base}/200`,
        },
        {
          id: "2",
          url: `${base}/200`,
        },
        {
          id: "3",
          url: `${base}/200?delay=100`,
        },
      ],
    });
    expect(res.bag.states["1"].pending).toEqual(true);
    expect(res.bag.states["2"].pending).toEqual(true);

    res.bag.abort("3");

    await waitUntil(() => !res.bag.states["1"].pending);
    await waitUntil(() => !res.bag.states["2"].pending);

    expect(res.bag.states["3"].pending).toEqual(false);
    expect(res.bag.states["3"].rejected).toEqual(true);
    expect(res.bag.states["3"].error).not.toBeUndefined()
  });
});
