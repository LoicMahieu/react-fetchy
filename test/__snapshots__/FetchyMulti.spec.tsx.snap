// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FetchyMulti basic children 1`] = `
Object {
  "abort": [Function],
  "retry": [Function],
  "states": Object {},
}
`;

exports[`FetchyMulti basic render 1`] = `
Object {
  "abort": [Function],
  "retry": [Function],
  "states": Object {},
}
`;

exports[`FetchyMulti fetch on mount 1`] = `
Object {
  "abort": [Function],
  "retry": [Function],
  "states": Object {
    "1": Object {
      "error": undefined,
      "fulfilled": true,
      "pending": false,
      "progress": undefined,
      "rejected": false,
      "result": Object {
        "header": Object {
          "connection": "close",
          "content-type": "application/json",
          "date": "Fake",
          "transfer-encoding": "chunked",
        },
        "req": Object {
          "data": undefined,
          "headers": Object {
            "some-header": "1",
            "user-agent": "node-superagent/3.8.3",
          },
          "method": "GET",
          "url": "http://localhost:8081/200?foo",
        },
        "status": 200,
        "text": "{\\"headers\\":{\\"host\\":\\"localhost:8081\\",\\"accept-encoding\\":\\"gzip, deflate\\",\\"user-agent\\":\\"node-superagent/3.8.3\\",\\"some-header\\":\\"1\\",\\"connection\\":\\"close\\"},\\"url\\":\\"/200?foo\\"}",
      },
      "value": Object {
        "headers": Object {
          "accept-encoding": "gzip, deflate",
          "connection": "close",
          "host": "localhost:8081",
          "some-header": "1",
          "user-agent": "node-superagent/3.8.3",
        },
        "url": "/200?foo",
      },
    },
    "2": Object {
      "error": undefined,
      "fulfilled": true,
      "pending": false,
      "progress": undefined,
      "rejected": false,
      "result": Object {
        "header": Object {
          "connection": "close",
          "content-type": "application/json",
          "date": "Fake",
          "transfer-encoding": "chunked",
        },
        "req": Object {
          "data": undefined,
          "headers": Object {
            "user-agent": "node-superagent/3.8.3",
          },
          "method": "GET",
          "url": "http://localhost:8081/200?delay=50",
        },
        "status": 200,
        "text": "{\\"headers\\":{\\"host\\":\\"localhost:8081\\",\\"accept-encoding\\":\\"gzip, deflate\\",\\"user-agent\\":\\"node-superagent/3.8.3\\",\\"connection\\":\\"close\\"},\\"url\\":\\"/200?delay=50\\"}",
      },
      "value": Object {
        "headers": Object {
          "accept-encoding": "gzip, deflate",
          "connection": "close",
          "host": "localhost:8081",
          "user-agent": "node-superagent/3.8.3",
        },
        "url": "/200?delay=50",
      },
    },
  },
}
`;

exports[`FetchyMulti handle errors 1`] = `
Object {
  "abort": [Function],
  "retry": [Function],
  "states": Object {
    "1": Object {
      "error": [Error: Internal Server Error],
      "fulfilled": false,
      "pending": false,
      "rejected": true,
      "result": Object {
        "header": Object {
          "connection": "close",
          "date": "Fake",
          "transfer-encoding": "chunked",
        },
        "req": Object {
          "data": undefined,
          "headers": Object {
            "user-agent": "node-superagent/3.8.3",
          },
          "method": "GET",
          "url": "http://localhost:8081/500",
        },
        "status": 500,
        "text": "500",
      },
      "value": null,
    },
  },
}
`;

exports[`FetchyMulti handle errors and retry 1`] = `
Object {
  "abort": [Function],
  "retry": [Function],
  "states": Object {
    "1": Object {
      "error": [Error: Internal Server Error],
      "fulfilled": false,
      "pending": false,
      "rejected": true,
      "result": Object {
        "header": Object {
          "connection": "close",
          "date": "Fake",
          "transfer-encoding": "chunked",
        },
        "req": Object {
          "data": undefined,
          "headers": Object {
            "user-agent": "node-superagent/3.8.3",
          },
          "method": "GET",
          "url": "http://localhost:8081/500",
        },
        "status": 500,
        "text": "500",
      },
      "value": null,
    },
  },
}
`;
